pkg/exchange/binance/stream.go

```diff
@@ -146,34 +163,63 @@ func (s *Stream) handleDisconnect() {
 }
 
 func (s *Stream) handleConnect() {
+	var err error
 	if !s.PublicOnly {
-		// Emit Auth before establishing the connection to prevent the caller from missing the Update data after
-		// creating the order.
+		if s.ed25519Auth {
+			err = s.authConnectEd25519()
+		} else {
+			log.Warnf("Ed25519 private key is not set, using deprecated HMAC authentication")
+			err = s.authConnectHMAC()
+		}
+	} else {
+		err = s.subscribe()
+	}
+	if err != nil {
+		log.WithError(err).Error("subscribe error")
+	}
+}
+
+func (s *Stream) authConnectHMAC() error {
+	go s.EmitAuth()
+	return nil
+}
 
-		// spawn a goroutine to emit auth event to prevent blocking the main event loop
-		go s.EmitAuth()
-		return
+func (s *Stream) authConnectEd25519() error {
+	timestamp := time.Now().UnixNano() / 1e6
+	params := map[string]string{
+		"apiKey":    s.client.APIKey,
+		"timestamp": fmt.Sprintf("%d", timestamp),
 	}
+	signature := api.GenerateSignatureEd25519(params, s.privateKey)
+	return s.Conn.WriteJSON(LoginCommand{
+		WebSocketCommand: WebSocketCommand{
+			Method: "session.login",
+			ID:     1,
+		},
+		Params: LoginParams{
+			APIKey:    s.client.APIKey,
+			Signature: signature,
+			Timestamp: timestamp,
+		},
+	})
+}
 
+func (s *Stream) subscribe() error {
```

Comment:
maybe name it `sendSubscribe` or `writeSubscribe`

Reply:
I have a better idea, how about `subscribeMarketDataSource`?
Since `subscribe` is only used for market data stream, naming it this way make it more clear in my opinion.

Reply:
Ok, `writeSubscriptions` it is!